example/particle_tracers/particle_tracers.cpp

```diff
@@ -0,0 +1,515 @@
+  }
+
+  return TaskStatus::complete;
+}
+
+TaskStatus Defrag(MeshBlock *pmb) {
+  auto s = pmb->swarm_data.Get()->Get("tracers");
+
+  // Only do this if list is getting too sparse. This criterion (whether there
+  // are *any* gaps in the list) is very aggressive
+  if (s->GetNumActive() <= s->GetMaxActiveIndex()) {
+    s->Defrag();
+  }
+
+  return TaskStatus::complete;
+}
+
+// Mark all MPI requests as NULL / initialize boundary flags.
+TaskStatus InitializeCommunicationMesh(const BlockList_t &blocks) {
+  // Boundary transfers on same MPI proc are blocking
+  for (auto &block : blocks) {
+    auto swarm = block->swarm_data.Get()->Get("tracers");
+    for (int n = 0; n < block->pbval->nneighbor; n++) {
+      NeighborBlock &nb = block->pbval->neighbor[n];
+      swarm->vbswarm->bd_var_.req_send[nb.bufid] = MPI_REQUEST_NULL;
+    }
+  }
+
+  for (auto &block : blocks) {
+    auto &pmb = block;
+    auto sc = pmb->swarm_data.Get();
+    auto swarm = sc->Get("tracers");
+
+    for (int n = 0; n < swarm->vbswarm->bd_var_.nbmax; n++) {
+      auto &nb = pmb->pbval->neighbor[n];
+      swarm->vbswarm->bd_var_.flag[nb.bufid] = BoundaryStatus::waiting;
+    }
+  }
+
+  // Reset boundary statuses
+  for (auto &block : blocks) {
+    auto &pmb = block;
+    auto sc = pmb->swarm_data.Get();
+    auto swarm = sc->Get("tracers");
+    for (int n = 0; n < swarm->vbswarm->bd_var_.nbmax; n++) {
+      auto &nb = pmb->pbval->neighbor[n];
+      swarm->vbswarm->bd_var_.flag[nb.bufid] = BoundaryStatus::waiting;
+    }
+  }
+
+  return TaskStatus::complete;
+}
+
+} // namespace Particles
+
+// *************************************************//
+// define the application driver. in this case,    *//
+// that just means defining the MakeTaskList       *//
+// function.                                       *//
+// *************************************************//
+
+void ProblemGenerator(MeshBlock *pmb, ParameterInput *pin) {
+  auto &pkg = pmb->packages.Get("particles_package");
+  auto &rc = pmb->meshblock_data.Get();
+  auto &density = rc->Get("density").data;
+  auto &swarm = pmb->swarm_data.Get()->Get("tracers");
+  const auto num_tracers = pkg->Param<int>("num_tracers");
+  auto rng_pool = pkg->Param<RNGPool>("rng_pool");
+
+  const IndexRange &ib = pmb->cellbounds.GetBoundsI(IndexDomain::interior);
+  const IndexRange &jb = pmb->cellbounds.GetBoundsJ(IndexDomain::interior);
+  const IndexRange &kb = pmb->cellbounds.GetBoundsK(IndexDomain::interior);
+  auto coords = pmb->coords;
+
+  const Real density_mean = 1.0;
+  const Real density_amp = 0.5;
+  PARTHENON_REQUIRE(density_mean > density_amp, "Cannot have negative densities!");
+
+  auto density_h = density.GetHostMirror();
+  for (int k = kb.s; k <= kb.e; k++) {
+    for (int j = jb.s; j <= jb.e; j++) {
+      for (int i = ib.s; i <= ib.e; i++) {
+        density_h(k, j, i) = density_mean + density_amp * sin(2. * M_PI * coords.x1v(i));
+      }
+    }
+  }
+  density.DeepCopy(density_h);
```

Comment:
Could we put this in a `par_for`? Just to avoid a host copy? I know it doesn't matter since it's initialization but it would still be nicer I think.

Reply:
Yeah that's better (also will `coords.x1v(i)` even work on the host if you compile for GPU? I assume not). I think there was going to be a reason this should be done on the host at some point, but there isn't now.

Reply:
coords should work on both host and device I think. At least for now.